docs/DevGuide/Databox.md

```diff
@@ -0,0 +1,366 @@
+address the desire to combine the functionality of tags with the modularity of
+functions, DataBox provides ComputeTags.
+
+## ComputeTags {#documentation_for_compute_tags}
+ComputeTags are used to tag functions that are used in conjunction with a
+DataBox to produce a new quantity. ComputeTags look like:
+
+\snippet Test_DataBoxDocumentation.cpp compute_tags
+
+ComputeTags derive off of `db::ComputeTag`, and it is convenient to have them
+additionally derive off of an existing SimpleTag (in this case `Mass`) so that
+the quantity `MassCompute` can be obtained through the SimpleTag `Mass`.
+We use the naming convention `TagNameCompute` so that `TagNameCompute` and
+`TagName` appear next to each other in documentation that lists tags in
+alphabetical order.
+
+We have also added the type alias `argument_tags`, which is necessary in order
+to refer to the correct tagged quantities in the DataBox.
+
+\note
+The `tmpl::list` used in the type alias is a non-associative container only
+holding types. That is, there is no variable runtime data associated with it
+like there is for `std::tuple`, which is a container associating types with
+values. They are useful in situations when one is working with multiple tags at
+once.
+
+This compile time resolution of *metavariables* is a pattern that is used again
+and again in SpECTRE. Let us see how we can compute our beloved quantity
+of mass times acceleration:
+
+\snippet Test_DataBoxDocumentation.cpp compute_tags_force_compute
+
+And that's it! `db::get` utilizes the `argument_tags` specified in
+`ForceCompute` to determine which items to get out of the `refined_databox`.
+With the corresponding quantities in hand, `db::get` passes them as arguments
+to the `function` specified in `ForceCompute`. This is why every `ComputeTag`
+must have an `argument_tags` as well as a `function` specified; this is the
+contract with DataBox they must satisfy in order to enjoy the full benefits of
+DataBox's generality.
+
+## Mutating DataBox items {#documentation_for_mutate_tags}
+It is reasonable to expect that in a complex calculation, we will encounter
+time-dependent or iteration-dependent variables. As a result, in addition to
+adding and retrieving items from our DataBox, we also need a way to *mutate*
+quantities already present in the DataBox. This can be done via
+`db::mutate_apply` and can look like the following:
+
+\note
+There is an alternative to `db::mutate_apply`, `db::mutate`. See the
+\ref DataBoxGroup "DataBox documentation" for more details.
+
+\snippet Test_DataBoxDocumentation.cpp mutate_tags
+\snippet Test_DataBoxDocumentation.cpp time_dep_databox
+
+\note
+The `not_null`s here are used to give us the assurance that the pointers
+`time` and `falling_speed` are not null pointers. Using raw pointers alone, we
+risk running into run time errors if we fail to perform additional checks.
+For information on the usage of `not_null`, see the documentation for Gsl.hpp.
```

Comment:
`not_null` gives a runtime error that tells the user what happened, while dereferencing a nullptr just gives a segfault.